core/tee/tee_supp_plugin_rpc.c

```diff
@@ -0,0 +1,70 @@
+// SPDX-License-Identifier: BSD-2-Clause
+/*
+ * Copyright (c) 2020, Open Mobile Platform LLC
+ */
+
+#include <assert.h>
+#include <kernel/thread.h>
+#include <mm/mobj.h>
+#include <optee_rpc_cmd.h>
+#include <stddef.h>
+#include <stdint.h>
+#include <stdlib.h>
+#include <string.h>
+#include <tee/tee_supp_plugin_rpc.h>
+#include <tee/uuid.h>
+#include <trace.h>
+
+TEE_Result tee_invoke_supp_plugin_rpc(const TEE_UUID *uuid, uint32_t cmd,
+				      uint32_t sub_cmd, void *buf, size_t len,
+				      size_t *outlen)
+{
+	TEE_Result res = TEE_ERROR_GENERIC;
+	struct thread_param params[THREAD_RPC_MAX_NUM_PARAMS];
+	uint32_t *plugin_uuid = NULL;
+	void *va = NULL;
+	struct mobj *mobj = NULL;
+
+	if (!uuid || (len && !buf) || (!len && buf))
+		return TEE_ERROR_BAD_PARAMETERS;
+
+	if (len) {
+		mobj = thread_rpc_alloc_payload(len);
+		if (!mobj) {
+			EMSG("can't create mobj for plugin data");
+			return TEE_ERROR_OUT_OF_MEMORY;
+		}
+
+		va = mobj_get_va(mobj, 0);
+		if (!va) {
+			EMSG("can't get va from mobj");
+			goto out;
+		}
+
+		memcpy(va, buf, len);
+	}
+
+	plugin_uuid = (uint32_t *)uuid;
```

Comment:
I believe you need to do something like:
``` C
uint32_t uuid_words[4] = { };
tee_uuid_to_octets(uuid_words, uuid);
```
Secure world is always in little endian so there's no need for any more byte swapping.

Later when decoding this in tee-supplicant you'd do something like:
``` C
TEEC_UUID uuid = { };
uint32_t uuid_words[4] = { };

uuid_words[0] = from_le32(params[0].b);
...
tee_uuid_from_octets(&uuid, (void *)uuid_words);
```
This should work for both a big and little endian normal world. Other parts in the chain (driver) are missing so a big endian normal world wouldn't work for other reasons, but we shouldn't make it harder than necessary in case we'd ever fix that.

Do it make sense?

Reply:
Yes, it make sense. I added `le32toh()` to tee-supplicant when decoding uuid_words[].